W-11862525-first-draft-for-beta-site-build-CP (#114)
mulesoft/docs-apikit: modules/ROOT/pages/validate-4-task.adoc

> = Validating Query Parameters and Headers

Reviewer: Check the guidelines for gerund/imperative headings. I believe this should be imperative.

Author: @hannanelson, thanks for your feedback!
I understand that we should use gerund in H1 and infinitive in H2+.
Actually, the APIkit beta site (https://dev-docs-internal.kqa.msap.io/beta-apikit-new-content-structure/apikit/4.x/overview) has been structured like that.
Please, can you confirm if gerund is OK for H1 and infinitive for H2+?
Thanks!

Reviewer: Yes. If there are multiple tasks, the H1 should be gerund and the H2s imperative.

Author: @hannanelson, in this case, we have task 1 = H1 and task 2 = H2. So we can consider we have two tasks.
As per our style guide, we should leave the gerund in this case:
"Task pages introduced using a gerund title and narrative text about an action typically lead to one or more imperative-titled task subtopics (Build Your Database, Create a Message Flow) include numbered steps."
Please, can you confirm?
Thanks!

ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

> Configure *APIkit Router* to enforce validation on headers and query parameters, restricting them only to the attributes specified within your API specification.

Reviewer: Can you put the user task first here?

Reviewer: Check the placement of only.
https://confluence.internal.salesforce.com/pages/viewpage.action?spaceKey=MTDT&title=MuleSoft+CX+Writing+Style+Reference#MuleSoftCXWritingStyleReference-%22only%22(%22simply,%22%22just%22)

Author: @hannanelson, this article explains the validation of query parameters and headers, and the paragraph explains this task. So sorry to say I don't get what you mean by putting the user task first.
Thanks for your support!

// tag::process[]
. In Studio, select *APIkit Router*.
+
image::apikit-router-selection.png["APIkit router highlighted"]
. Click *Edit* for your router configuration.
+
image::router-config.png["Edit button highlighted in the APIkit router tab"]
// end::process[]

> . Select *Query Parameters Strict Validations* to restrict the usage of query parameters to only the ones defined in your API specification. 

Reviewer: Suggest something like:
Specify validation settings:
- Query Parameters Strict Validations
- Headers Strict Validations

Author: Thanks for your comment, @hannanelson! 
Your option is shorter and I like it! Nevertheless, it may be confusing for the readers because the explanation lacks the purpose of the configurations, i.e.: to restrict the usage of XXX to the ones defined in the users' API specification.
Or did I misunderstand your comment?
Thanks!

Reviewer: I didn't mean to remove the descriptions, just change the format of the list to add a sub-bullet list.

Author: I may be misunderstanding your point. But there is no list in this file. There are steps. And each step provides the step itself and its purpose.
Thanks!

> . Select *Headers Strict Validations* to limit the headers to only the ones specified in your API specification.

Reviewer: I think limit and only are redundant here.
Suggest something like "to limit headers to those specified in your API specification."

Author: You're right!
Reworded! Thanks!

+

> image::enable-verifications.png["*Query parameters Strict Validations* and *Headers Strict Validations* highlighted in the router configuration tab"]

Reviewer: Do we need formatting in ALT text? I'm not sure they appear.

Author: I'm just formatting ALT text because it is easier to read. Without the asterisks, the two noun phrases may be hard to find. Please, let me know if formatting negatively affects ALT texts. 
Thanks!

Reviewer: I don't know the answer to this question. Perhaps @cheungaryk  can help?

> If not, update the APIkit configuration code as follows:

Reviewer: If not what?

Author: requesting assistance from SMEs.
Thanks!

Author: There are two ways to achieve the same result: following the mentioned steps in Studio UI or changing the XML code.
Sentences were rewritten. Thanks!

[source,xml,linenums]
----
<apikit:config  name="api-config"
                api="api.raml"
                outboundHeadersMapName="outboundHeaders"
                httpStatusVarName="httpStatus"
                queryParamsStrictValidation="true"
                headersStrictValidation="true" />
----

== Disable Validation of Query Parameters and Headers

> If your specific scenario and use case can handle it, disable the validation of query parameters and headers to minimize latency.

Reviewer: What's the difference between scenario and use case?
How do I know if they can handle it? What is "it" in this case?

What about something like: "To minimize latency, disable..."

Author: I like your idea! 
If we can not specify the context, the "specific scenario and use case can handle it" may mean nothing to the readers.
Thanks!

include::page$validate-4-task.adoc[tag=process]
. Select *Disable Validations*.
+
image::disable-verifications-apikit.png["*Disable validations* option highlighted in the router configuration tab"]

> If not, update the APIkit configuration code as follows:

Reviewer: What does "If not" mean here? If I don't select Disable Validations? If I don't do it via the UI?

Author: Requesting assistance from SMEs.

Author: There are two ways to achieve the same result: following the mentioned steps in Studio UI or changing the XML code.
Sentences were rewritten. Thanks!

+
image::router-config.png["Edit button highlighted next to the plus button, in the APIkit router tab."]
// end::process[]
. Enable *Query Parameters Strict Validations* to restrict the usage of query parameters solely to those that are defined in your API specification. Similarly, enable *Headers Strict Validations* to limit the headers to only those specified in your API specification:
+
image::enable-verifications.png["Query parameters strict validations and headers strict validations highlighted in the router configuration tab."]

Or just update the APIkit configuration code as following:

[source,xml,linenums]
----
<apikit:config  name="api-config"
                api="api.raml"
                outboundHeadersMapName="outboundHeaders"
                httpStatusVarName="httpStatus"
                queryParamsStrictValidation="true"
                headersStrictValidation="true" />
----

== Disable Validation of Query Parameters and Headers

> If minimizing latency is of utmost importance for your specific scenario and your use case can tolerate the absence of validations, it is possible to deactivate them.

Reviewer 2: ```suggestion
If minimizing latency is important for your specific scenario and your use case can tolerate the absence of validations, you can deactivate them.
```

Author: OK! Thanks!
